Better error message for query utxo on oops

newhoggy opened the pull request:
This PR introduces the use of the `oops` library for simplifying error handling.  Functions that use `oops` for error handling end in an underscore.  For example:

* `determineEraExpr_`
* `getNtcVersion_`
* `getSbeInQuery_`
* `handleQueryConvenienceErrors_`
* `maybeQueryExpr_`
* `queryEraHistory_`
* `queryExpr_`
* `queryNodeLocalState_`
* `queryProtocolParams_`
* `queryStakePools_`
* `queryStateForBalancedTx_`
* `querySystemStart_`
* `queryUtxo_`
* `setupLocalStateQueryExpr_`
* `writeStakePools_`

In most cases, equivalent functions that don't use `oops` types in their type signature are retained, but have been re-implemented to call the `oops` version of the function.  This is to allow code to gradually migrate to the `oops` error handling style and give time to learn the new style.

As a result of these changes `executeQueryCardanoMode` and `executeQueryAnyMode` become unused and are removed.

# Reading notes

This PR adds code that uses a different error handling style supported by the `oops` library.  Functions that have a type `oops` enabled signature end in an underscore so that they are easily distinguishable from others.

Use of `BlockArguments` has been removed from the PR because it was optional in the first place and there was too much happening in the one PR.

# Stylistic choices

As part of implementing this feature, select parts of the code base has been converted to use `oops` error handling library to solve a particularly nasty problem.

The shape of the problem can be described like this:

> We have a monadic type called `LocalStateQueryExpr` which is used to express local state queries.
>
> This monad was introduced to allow complex expressions to be written.  These expressions can contain multiple queries.
> 
> Packaging multiple queries into the one expression means that the CLI can perform multiple interdependent queries over the same connection.  Prior to the introduction of this monadic type, the only way to implement a CLI command that required multiple queries was to connect to the validator node multiple times once per query.
>
> The new function `executeLocalStateQueryExpr_` acts as the interface between the `IO` monad and this query expression monad.
>
> The new monadic type introduces new problems of its own.  For example `executeLocalStateQueryExpr_` can fail with an `AcquireFailure`.  The possibility of this error is imposed by the networking library we use.
>
> Moreover, some, but not all query expressions may fail with `UnsupportedNtcVersionError`.  This is the feature that this PR implements.
>
> Further to the problem is that a query expression written by the developer may additionally want to fail with its own error.
>
> The difficulty with multiples parts of the same system wanting to throw their own errors is that the `executeLocalStateQueryExpr_` must now accomodate errors raised by each of these possibilities.
>
> The simplest solution might have been to return a stack of `Either`s, but large stacks of `Either`s are unwieldly.  I would have to worry about the ordering of the errors in the stack and carefully unpack the error I want to handle (which could be nested) and rebuild the rest of the `Either` stack that I don't want to handle just yet so those errors can be popagated.
>
> Without the help of a good error handling library that allows multiple errors to be thrown generically and transparently across a function boundary, the burden of juggling multiple error types is imposed on the user.
>
> Attempts have been made to solve this problem some other way.   For example `executeLocalStateQueryExpr_` at one stage took an additional argument that served as a mapping function from errors that `executeLocalStateQueryExpr_` need to throw to some generic error that the query expression could throw.
>
> The problem was further compounded by having to support `UnsupportedNtcVersionError` which means that often the developer will need to map the error twice to cross two function boundaries.
>
> All of this means that it is often too difficult to write monadic queries and developers hardly ever write them and instead just run queries over separate connections.

The problem is a long lived one I've been trying to solve since July 2021 in this PR https://github.com/input-output-hk/cardano-node/pull/2957

Using the `oops` library allows us to neatly solve the above problem.  For example the `executeLocalStateQueryExpr_` is now:

```
executeLocalStateQueryExpr_
  :: forall e mode a . ()
  => e `OO.CouldBe` AcquireFailure
  => LocalNodeConnectInfo mode
  -> Maybe ChainPoint
  -> (NodeToClientVersion -> ExceptT (OO.Variant e) (LocalStateQueryExpr (BlockInMode mode) ChainPoint (QueryInMode mode) () IO) a)
  -> ExceptT (OO.Variant e) IO a
executeLocalStateQueryExpr_ connectInfo mpoint f = ...
```

In the above, `e` is a type list of all possible types that could be thrown.

There is a constraint `CouldBe` constraint that says that `AcquireFailure` must at least in the type list.

This means that `executeLocalStateQueryExpr_` is able to throw `AcquireFailure` while the query expression in function `f` is free to throw whatever it wants.

To show how functions using `oops` are implemented, but `executeQuery` function is reproduced below:

```
executeQuery
  :: forall result era mode. CardanoEra era
  -> ConsensusModeParams mode
  -> LocalNodeConnectInfo mode
  -> QueryInMode mode (Either EraMismatch result)
  -> ExceptT ShelleyQueryCmdError IO result
executeQuery era cModeP localNodeConnInfo q = do
  eraInMode <- calcEraInMode era $ consensusModeOnly cModeP
  case eraInMode of
    ByronEraInByronMode -> left ShelleyQueryCmdByronEra
    _ -> OO.runOops1 @ShelleyQueryCmdError do
      executeLocalStateQueryExpr_ localNodeConnInfo Nothing
        do \_ -> do
            result <- queryExpr q
            case result of
              Right a -> return a
              Left e -> OO.throw (ShelleyQueryCmdLocalStateQueryError $ EraMismatchError e)
        & do OO.catch @AcquireFailure \e -> OO.throw $ ShelleyQueryCmdAcquireFailure $ toAcquiringFailure e
        & do OO.catch @UnsupportedNtcVersionError \e -> OO.throw $ ShelleyQueryCmdUnsupportedNtcVersion e
```

The first thing to note that whilst this function uses `oops`, its type signature is just `ExceptT ShelleyQueryCmdError`.

The interface between code that uses `oops` and regular `ExceptT` code is demarcated by `runOops1 @ShelleyQueryCommandError`.

This function embeds an `ExceptT (Variant '[x])` expression into a `ExceptT x` expression.

This means the `oops` expression in the following `do` block may only throw `ShelleyQueryCmdError`.

If functions inside this `do` block throw something else, they must be caught to return a value or re-throw as `ShelleyQueryCmdError`.

The catch and re-throw is done by the lines:

```
        & do OO.catch @AcquireFailure \e -> OO.throw $ ShelleyQueryCmdAcquireFailure $ toAcquiringFailure e
        & do OO.catch @UnsupportedNtcVersionError \e -> OO.throw $ ShelleyQueryCmdUnsupportedNtcVersion e
```

Failing to catch an except that may not be thrown results in an error that looks like this:

```
    • Uh oh! I couldn't find UnsupportedNtcVersionError inside the variant!
      If you're pretty sure I'm wrong, perhaps the variant type is ambiguous;
      could you add some annotations?
```

This is a sign that the developer must do something with `UnsupportedNtcVersionError`.  This can be one of the following options:

* Catch and re-throw another error that can be thrown.  In this case `ShelleyQueryCmdError`.
* Catch and return a value instead.
* Adjust the type signature of the surrounding function to allow propagation of `UnsupportedNtcVersionError` by adding `CouldBe e UnsupportedNtcVersionError`.  This is not an option in this case because `runOops1` imposes the restriction that only one error `ShelleyQueryCmdAcquireFailure` may propagate, but in other contexts, this may be possible.

An interesting feature of the `executeQuery` is that it makes heavy use of the `BlockArguments` Haskell extension:

```
    _ -> OO.runOops1 @ShelleyQueryCmdError do
      executeLocalStateQueryExpr_ localNodeConnInfo Nothing
        do \_ -> do
            result <- queryExpr q
            case result of
              Right a -> return a
              Left e -> OO.throw (ShelleyQueryCmdLocalStateQueryError $ EraMismatchError e)
        & do OO.catch @AcquireFailure \e -> OO.throw $ ShelleyQueryCmdAcquireFailure $ toAcquiringFailure e
        & do OO.catch @UnsupportedNtcVersionError \e -> OO.throw $ ShelleyQueryCmdUnsupportedNtcVersion e
```

Firstly the extension allows us to write `do` instead of `$ do`.

Secondly the expression that follows a `do` does not need to be a `Monad` or `Applicative`.  It can be a value of any type.

This happens here: `do \_ -> ...`.

In this way the `do` is kind of acting like a parentheses around the expression.

The expression is equivalent to `(\_ -> ...)`.

The contents of these virtual parentheses is determined by indent.

The second `do` in `do \_ -> do` is a regular `do` block.

Finally there is `& do OO.catch ...`.

Again, this is like `& (OO.catch ...)`.

```
executeQuery
  :: forall result era mode. CardanoEra era
  -> ConsensusModeParams mode
  -> LocalNodeConnectInfo mode
  -> QueryInMode mode (Either EraMismatch result)
  -> ExceptT ShelleyQueryCmdError IO result
executeQuery era cModeP localNodeConnInfo q = do
  eraInMode <- calcEraInMode era $ consensusModeOnly cModeP
  case eraInMode of
    ByronEraInByronMode -> left ShelleyQueryCmdByronEra
    _ -> OO.runOops1 @ShelleyQueryCmdError do
      ( executeLocalStateQueryExpr_ localNodeConnInfo Nothing
          (\_ -> do
              result <- queryExpr q
              case result of
                Right a -> return a
                Left e -> OO.throw (ShelleyQueryCmdLocalStateQueryError $ EraMismatchError e)))
        & (OO.catch @AcquireFailure \e -> OO.throw $ ShelleyQueryCmdAcquireFailure $ toAcquiringFailure e)
        & (OO.catch @UnsupportedNtcVersionError \e -> OO.throw $ ShelleyQueryCmdUnsupportedNtcVersion e)
```

`BlockArguments` allows us to not use parentheses which means the developer doesn't have to worry about closing parentheses lining up.  They can look at indent instead which is easier to see and typical of other Haskell code.

An added benefit is that fewer lines are touched when expressions need to be regrouped.

For example the opening parenthesis for the expression `Left e -> OO.throw (ShelleyQueryCmdLocalStateQueryError $ EraMismatchError e)))` are distributed across three different lines so regrouping at any of the three levels involves touching this line.

Additionally, note that the type signature of `catch` is not typical of catch functions.

Catch functions typically take the handler in the second argument.  `catch` however takes the handler in the first argument.

This is important because it means adding and removing handlers in the above code touches the fewest lines.  We don't have to worry about nesting expressions further (which can touch a lot of lines and cause large `git` diffs) or breaking out handlers into a where clause.

**The use of `BlockArguments` is entirely optional**.  However when evaluating its merits it is worth considering that Haskell provides two mechanisms for grouping expressions: parentheses and `do` blocks.

When refactoring code that has a mixture of both, the developer will have to maintain _both_ the indent and the parentheses and ensure they play well together.  `BlockArguments` allows the developer to no longer have to think about parentheses and just think about grouping code with indents.  This halves the cognitive burden that grouping of expressions imposes on the developer.

Finally, `oops` has decent type inference.

For example GHC and HLS can infer that the type of this expression:

```
queryExpr_ q = do
  let minNtcVersion = ntcVersionOf q
  ntcVersion <- getNtcVersion
  if ntcVersion >= minNtcVersion
    then lift
      $ LocalStateQueryExpr $ ReaderT $ \_ -> ContT $ \f -> pure $
        Net.Query.SendMsgQuery q $
          Net.Query.ClientStQuerying
          { Net.Query.recvMsgResult = f
          }
    else OO.throw $ UnsupportedNtcVersionError minNtcVersion ntcVersion
```

is this:

```
queryExpr_ :: OO.CouldBeF e UnsupportedNtcVersionError =>
                   QueryInMode mode b
                   -> ExceptT
                        (OO.Variant e)
                        (LocalStateQueryExpr block point (QueryInMode mode) r IO)
                        b
```

Which I can easily rewrite to my preferred type signature:

```
queryExpr_ :: ()
  => e `OO.CouldBe` UnsupportedNtcVersionError
  => QueryInMode mode a
  -> ExceptT (OO.Variant e) (LocalStateQueryExpr block point (QueryInMode mode) r IO) a
```

## Why not use another library

I have evaluated a number of other options.  `plucky`, `fused-effects`, `control-monad-exceptions`, `polysemy` and `haskus-utils-variant`.

They all suffer from a common problem which is the catch functions take the handler in the second argument making stacking handlers more fiddly that it needs to be.

Additionally they can have bad error messages.  Take for example `plucky`:

```
src/Cardano/CLI/Shelley/Run/Query.hs:1367:7: error:
    • No (remind me to write a better error message)
    • In the expression:
        executeLocalStateQueryExpr_ localNodeConnInfo Nothing
      In the second argument of ‘($)’, namely
        ‘executeLocalStateQueryExpr_ localNodeConnInfo Nothing
           $ \ _
               -> do result <- queryExpr q
                     case result of
                       Right a -> ...
                       Left e -> ...’
      In the expression:
        wrapExceptT ShelleyQueryCmdUnsupportedNtcVersion
          $ executeLocalStateQueryExpr_ localNodeConnInfo Nothing
              $ \ _
                  -> do result <- queryExpr q
                        case result of
                          Right a -> ...
                          Left e -> ...
     |
1367 |       executeLocalStateQueryExpr_ localNodeConnInfo Nothing $ \_ -> do
     |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Some libraries like `fused-effects` and `polysemy` make it difficult to interoperate with the rest of the Haskell ecosystem.  For example it is difficult to get `ResourceT` working if at all and a lot of libraries in the ecosystem uses `ResourceT`.

Other common interoperability pain points include callbacks (for example with FFI callbacks), lazy `IO` and `UnliftIO`.

`oops`, likely `plucky` is based on `ExceptT` which is common in the Haskell ecosystem and most developers know how to use `ExceptT`.  `oops` imposes very little friction when interoperating with libraries that don't use `oops`.

In some cases, a library's type inference is not up to scratch.  For example in `oops`, if I try to infer the type of the expression:

```
foo_ = OO.throw () & OO.catch (\() -> pure ())
```

I get this:

```
foo_ :: ExceptT (OO.Variant e') ghc-prim-0.6.1:GHC.Types.Any ()
```

Which isn't quite right, but I can guess what the type signature should be:

```
foo_ :: Monad m => ExceptT (OO.Variant e') m ()
```

In `plucky` I'm not actually sure what the type of this is:

```
foo = PKY.throwT () & PKY.catchT (\() -> pure ())
```

The type doesn't compile to start with and the inferred type is wrong:

```
foo :: ExceptT e' m a
```

I would expect the type to be something like this to indicate no exceptions are thrown:

```
foo :: Monad m => ExceptT Void m a
```

but that doesn't work.

The following doesn't work either:

```
foo :: Monad m => ExceptT e m ()
```

The example might seem academic, but it is not.  If `oops` fails to infer a valid type (is there even one) when the code catches all errors, then a developer could easily get into this situation by catching all errors.  Code with all errors caught is valid code so it ought to have a type and that type should be inferrable.

Finally `oops` does one more thing to help the developer past type inference issues.  Every function that is used for error handling has an explicit `forall` and the the generic type argument for the error `x` is always the first argument.  This is because when writing error handling code where there may be multiple possible error types at place it is very beneficial for the developer to be able to use `TypeApplications` to fix the error type to a particular type.  Doing this will often resolve type inference problems.

newhoggy commented on cardano-api/src/Cardano/Api/Convenience/Query.hs:
```diff
@@ -62,59 +66,46 @@ queryStateForBalancedTx
   -> NetworkId
   -> [TxIn]
-  -> IO (Either QueryConvenienceError (UTxO era, ProtocolParameters, EraHistory CardanoMode, SystemStart, Set PoolId))
+  -> ExceptT QueryConvenienceError IO (UTxO era, ProtocolParameters, EraHistory CardanoMode, SystemStart, Set PoolId)
```
I think possible this happened because this came from a very old PR that got rebase (original from July 2021).  I will fix this up.

newhoggy replied:
Done.

newhoggy commented on cardano-api/src/Cardano/Api/Query.hs:
```diff
@@ -263,6 +277,23 @@ data QueryInShelleyBasedEra era result where
     :: PoolId
     -> QueryInShelleyBasedEra era (SerialisedStakeSnapshots era)
 
+instance NtcVersionOf (QueryInShelleyBasedEra era result) where
```
The `NodeToClientVersion` is never used by the users of the `API`.  It is used internally to map from the query to which version of the protocol supports it.

Users will call `queryExpr` and pass to it a `LocalStateQueryExpr`.  If any of the queries used in that expression correspond to an NTC version that isn't supported by the node a `UnsupportedNtcVersionError` is returned.

abailly-iohk reviewed (commented):
It's hard to comment on precise changes without having a deep knowledge of the code, which I don't have, but here are my 2 cts:
1. Being unfamiliar with the `BlockArguments` extension makes the code harder to read to me. Maybe it's only me, maybe it's not uncommon in which case we've made potential external contributors' work harder
2. I fail to see how the pattern `OO.catchM @AcquireFailure \e -> OO.throwM $ TxGenError $ show e` is better than using `catches` with a list of `Handle a`
3. This code adds quite a few layers (Oops monad, exceptT monad) which adds more burden to the potential future person wanting to refactor this or add a new feature
4. This requires wrapping things in `ExceptT` and at the same time use exceptions which seems somewhat contradictory to me: I would expect a strategy for error handling either embraces exceptions or explicit return types but not both for the same layer
5. This PR introduces heavy changes to code used quite often without a single test being changed, which would make me nervous if I were a maintainer of this codebase
6. As already discussed on slack, this code adds yet another dependency on an external library which implies new contributors would need to know and understand this library

abailly-iohk commented:
As an additional comment: I unfortunately don't have the time to provide some counterproposal that would solve the problem while IMO being simpler and easier to maintain, which I regret and implies you should take my comments with a grain of salt.

newhoggy commented:
> This requires wrapping things in ExceptT and at the same time use exceptions which seems somewhat contradictory to me: I would expect a strategy for error handling either embraces exceptions or explicit return types but not both for the same layer.

The PR as it currently stands uses a lot of `ExceptT` wrapping, but it does so because it interfaces with code with existing code.  The use of the wrapping shows how that interoperating with existing code is possible and there is no the need to rewrite the entire code base in one go.  As more of the code transitions to the proposed style the `ExceptT` wrapping would disappear, as would a lot of the error handling because errors could propagate which is a thing that doesn't happen when using `Either` and rarely happens with `ExceptT`.

newhoggy commented:
Add more notes on why `oops` in this comment https://github.com/input-output-hk/cardano-node/pull/4788#discussion_r1071753487

newhoggy commented on cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs:
```diff
@@ -1023,42 +1034,44 @@ runQueryStakePools
-                          network mOutFile = do
-  SocketPath sockPath <- firstExceptT ShelleyQueryCmdEnvVarSocketErr
-                           $ newExceptT readEnvSocketPath
+runQueryStakePools (AnyConsensusModeParams cModeParams) network mOutFile = OO.runOops1 @ShelleyQueryCmdError $ do
```
How `runQueryStakePools` is implemented using `oops`.

newhoggy replied:
For comparison https://github.com/input-output-hk/cardano-node/pull/4788/files#r1071976898

newhoggy replied:
With `oops`:

```
runQueryStakePools
  :: AnyConsensusModeParams
  -> NetworkId
  -> Maybe OutputFile
  -> ExceptT ShelleyQueryCmdError IO ()
runQueryStakePools (AnyConsensusModeParams cModeParams) network mOutFile = OO.runOops1 @ShelleyQueryCmdError $ do
  SocketPath sockPath <- liftIO readEnvSocketPath
    & OO.leftM @EnvSocketError (\e -> OO.throwM $ ShelleyQueryCmdEnvVarSocketErr e)

  let localNodeConnInfo = LocalNodeConnectInfo cModeParams network sockPath

  result <- executeLocalStateQueryExpr_ localNodeConnInfo Nothing
    ( do
        anyE@(AnyCardanoEra era) <- case consensusModeOnly cModeParams of
          ByronMode -> return $ AnyCardanoEra ByronEra
          ShelleyMode -> return $ AnyCardanoEra ShelleyEra
          CardanoMode -> queryExpr_ $ QueryCurrentEra CardanoModeIsMultiEra

        let cMode = consensusModeOnly cModeParams

        eInMode <- toEraInMode era cMode
          & OO.throwNothingAsM (ShelleyQueryCmdEraConsensusModeMismatch (AnyConsensusMode cMode) anyE)

        sbe <- getSbeInQuery_ $ cardanoEraStyle era

        (queryExpr_ $ QueryInEra eInMode $ QueryInShelleyBasedEra sbe $ QueryStakePools)
          & OO.leftM (\e -> OO.throwM (ShelleyQueryCmdEraMismatch e)))

    & OO.catchM @AcquireFailure (\e -> OO.throwM $ ShelleyQueryCmdAcquireFailure $ toAcquiringFailure e)
    & OO.catchM @UnsupportedNtcVersionError (\e -> OO.throwM $ ShelleyQueryCmdUnsupportedNtcVersion e)

  writeStakePools_ mOutFile result
```

newhoggy replied:
I've simplified the `oops` library:

* The combinator names no longer end in `M`
* Adopted familiar function names like `hoistMaybe` that are found in other libraries
* Taken in account discussion from other PRs regarding naming

For more information, see https://github.com/haskell-works/oops/pull/11

newhoggy commented on cardano-api/src/Cardano/Api/Convenience/Query.hs:
```diff
@@ -62,59 +66,46 @@ queryStateForBalancedTx
+    Nothing -> throwE $ EraConsensusModeMismatch (AnyConsensusMode CardanoMode) (getIsCardanoEraConstraint era $ AnyCardanoEra era)
 
 -- | Query the node to determine which era it is in.
 determineEra
```
Fixed.

newhoggy commented on cardano-api/src/Cardano/Api/IPC/Monad.hs:
```diff
@@ -45,69 +55,99 @@ import           Cardano.Api.Modes
+-- | Lift a query value into a monadic query expression.
+-- Use 'queryExpr' in a do block to construct monadic local state queries.
+queryExpr :: ()
+  => ProjectError e UnsupportedNtcVersionError
```
The commit related to this comment is gone.

newhoggy commented on cardano-api/src/Cardano/Api/Convenience/Query.hs:
```diff
@@ -58,45 +70,100 @@ renderQueryConvenienceError ByronEraNotSupported =
+    stakePools <- queryStakePools qeInMode qSbe
 
-  return (utxo, pparams, eraHistory, systemStart, stakePools)
+    pure (utxo, pparams, eraHistory, systemStart, stakePools)
```
Using `oops`, the query expression is very clean.  Individual queries within the expression can throw any number of errors and without requiring the rewriting of other query handling.

This contrasts to the `ExceptT`/`Either` only solution:  https://github.com/input-output-hk/cardano-node/pull/4825/files#r1083338808

newhoggy commented on cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs:
```diff
@@ -429,9 +429,25 @@ runTxBuildCmd
+                        systemStart <- querySystemStart_
+
+                        pure (utxo, eraHistory, systemStart)
+                    ) & handleQueryConvenienceErrors_
```
Inline `queryStateForBalancedTx_` to avoid unnecessarily querying for `pparams` and `stakePools` which are never used.

newhoggy commented on cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs:
```diff
@@ -429,9 +429,25 @@ runTxBuildCmd
-                      & handleQueryConvenienceErrors_
+                  (nodeEraUTxO, eraHistory, systemStart) <-
+                    (do
+                      let cModeParams2 = CardanoModeParams $ EpochSlots 21600
```
Is there any reason to not use `cModeParams`?

Jimbo4350 replied:
No it should be fine. Did it not compile when you tried to use the `localNodeConnInfo` from the higher scope?

newhoggy replied:
It compiles.  Updated to use the `localNodeConnInfo` from the the higher scope.

MarcFontaine commented:
Looks like total overkill to me.
I have the feeling that there is a lot of low hanging fruit to pick
which would help to make the code cleaner and more readable,
before jumping for the `oops` library.
The `oops` library seems to be brand new. Has it been used in any major Haskell project before.

Jimbo4350 reviewed (changes requested):
I'll have another pass at this next week when I'm back. I'm in agreement with @abailly-iohk's comment. I will dedicate time into this to be sure if `oops` is really the way to go or if there are existing abstractions we can use to simplify the code.

newhoggy commented on cardano-api/src/Cardano/Api/Convenience/Query.hs:
```diff
@@ -1,20 +1,28 @@
 
+import           Prelude
+
+import qualified Control.Monad.Oops as OO
```
The things I'm concerned about are things like `onLeft`, `onNothing, `throw`, and `catch` which conflict with other libraries.

newhoggy commented:
> I have the feeling that there is a lot of low hanging fruit to pick which would help to make the code cleaner and more readable, before jumping for the oops library.

This change is about improving error handling.  We don't have many good alternatives for improving error handling for the code we want to be able to write (multiple network queries over the same connection).

Until now, we've mostly avoided complicated error handling by avoiding writing queries that send multiple network queries over the same connection by open a new connection for each query, which is inefficient.

Recently I've made some PRs for improving error handing without `oops`, but this only takes us so far and doesn't actually solve the problem.

See https://github.com/input-output-hk/cardano-node/pull/4825

> The oops library seems to be brand new. Has it been used in any major Haskell project before.

The library is fairly new on hackage, but it's based on work an unpublished library that has existed since `2019`: https://github.com/i-am-tom/oops

I chose this library over [plucky](https://hackage.haskell.org/package/plucky) another similar library that existed on Hackage since 2020 because `oops` produced better error messages and does type-inference better.  I determined this by trying out both libraries.

About `15` months ago tried `plucky` in this PR and have was not satisfied with the result: https://github.com/input-output-hk/cardano-node/pull/3390

The problem that we avoid writing multiple query over single connection code has existed for at least this long and we ought to make doing this as easy as writing single query over a single connection.

Unless there is a viable alternative proposal to solve this problem, I put forward that `oops` is the most reasonable option we have.

Jimbo4350 reviewed (dismissed):
This is also a decrease in readability/increase in complexity for the sake of handling errors that users should take steps to avoid, namely:
- Using up to date `cardano-cli` and `cardano-node`  versions to avoid version mismatch errors
- Querying in the appropriate mode i.e `CardanoMode`. It is extremely unlikely users will be running Byron or Shelley only modes and also mismatching their queries by specifying the wrong mode in `cardano-cli`. We default to `CardanoMode` in the cli which the majority, if not all users will be using.


I'll open a slack thread today on this.

newhoggy commented:
I've cleaned up the PR considerable.  The PR had undergone many rebases and it accumulated some code that wasn't relevant anymore.  I've also updated it to only use the qualified `OO.` prefix where necessary and grouped related changes together into the same commit and reduced the size of each commit to easy reviewing.

newhoggy commented on cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs:
```diff
@@ -302,23 +306,27 @@ runQueryTip (AnyConsensusModeParams cModeParams) network mOutFile = do
+              eraHistory <- queryExpr_ (QueryEraHistory CardanoModeIsMultiEra)
+              mChainBlockNo <- queryExpr_ QueryChainBlockNo & OO.catchAsNothing @UnsupportedNtcVersionError
+              mChainPoint <- queryExpr_ (QueryChainPoint CardanoMode) & OO.catchAsNothing @UnsupportedNtcVersionError
+              mSystemStart <- queryExpr_ QuerySystemStart & OO.catchAsNothing @UnsupportedNtcVersionError
```
Example of error handling.

newhoggy commented on cardano-api/src/Cardano/Api/Convenience/Query.hs:
```diff
@@ -66,49 +145,66 @@ queryStateForBalancedTx
+    stakePools <- queryStakePools_ qeInMode qSbe
 
-  return (utxo, pparams, eraHistory, systemStart, stakePools)
+    pure (utxo, pparams, eraHistory, systemStart, stakePools)
```
Example of error handling.

carbolymer commented on cardano-cli/src/Cardano/CLI/Shelley/Run/Query.hs:
```diff
@@ -284,23 +289,27 @@ runQueryTip mNodeSocketPath (AnyConsensusModeParams cModeParams) network mOutFil
   case consensusModeOnly cModeParams of
     CardanoMode -> do
       let localNodeConnInfo = LocalNodeConnectInfo cModeParams network sockPath
+      eLocalState <- liftIO $ runExceptT $ runOopsInExceptT @ShelleyQueryCmdError $ do
```
isn't
```
runExceptT $ runOopsInExceptT == runOopsInEither
``` 
?

newhoggy replied:
Yep.  That's right.  Fixed!
